airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceOperations.java

```diff
@@ -49,12 +70,106 @@ public JsonNode rowToJson(final ResultSet queryContext) throws SQLException {
+        case BLOB, BINARY, VARBINARY, LONGVARBINARY -> putBinary(json, columnName, resultSet, colIndex);
+        default -> putDefault(json, columnName, resultSet, colIndex);
+      }
+    }
+  }
+
+  @Override
+  public JDBCType getFieldType(final JsonNode field) {
+    try {
+      final String typeName = field.get(INTERNAL_COLUMN_TYPE_NAME).asText();
+      // Postgres boolean is mapped to JDBCType.BIT, but should be BOOLEAN
+      if (typeName.equalsIgnoreCase("bool") || typeName.equalsIgnoreCase("boolean")) {
+        return JDBCType.BOOLEAN;
+      } else if (typeName.equalsIgnoreCase("bytea")) {
+        // BYTEA is variable length binary string with hex output format by default (e.g. "\x6b707a").
+        // It should not be converted to base64 binary string. So it is represented as JDBC VARCHAR.
+        // https://www.postgresql.org/docs/14/datatype-binary.html
+        return JDBCType.VARCHAR;
+      }
+
+      return JDBCType.valueOf(field.get(INTERNAL_COLUMN_TYPE).asInt());
+    } catch (final IllegalArgumentException ex) {
+      LOGGER.warn(String.format("Could not convert column: %s from table: %s.%s with type: %s. Casting to VARCHAR.",
+          field.get(INTERNAL_COLUMN_NAME),
+          field.get(INTERNAL_SCHEMA_NAME),
+          field.get(INTERNAL_TABLE_NAME),
+          field.get(INTERNAL_COLUMN_TYPE)));
+      return JDBCType.VARCHAR;
+    }
+  }
+
+  @Override
+  public JsonSchemaPrimitive getJsonType(final JDBCType jdbcType) {
+    return switch (jdbcType) {
+      case BOOLEAN -> JsonSchemaPrimitive.BOOLEAN;
+      case TINYINT, SMALLINT, INTEGER, BIGINT, FLOAT, DOUBLE, REAL, NUMERIC, DECIMAL -> JsonSchemaPrimitive.NUMBER;
+      case BLOB, BINARY, VARBINARY, LONGVARBINARY -> JsonSchemaPrimitive.STRING_BINARY;
+      default -> JsonSchemaPrimitive.STRING;
+    };
+  }
+
+  protected void putBoolean(final ObjectNode node, final String columnName, final ResultSet resultSet, final int index) throws SQLException {
+    node.put(columnName, resultSet.getString(index).equalsIgnoreCase("t"));
+  }
+
+  protected void putBigDecimal(final ObjectNode node, final String columnName, final ResultSet resultSet, final int index) throws SQLException {
+    final BigDecimal bigDecimal = DataTypeUtils.returnNullIfInvalid(() -> resultSet.getBigDecimal(index));
+    if (bigDecimal != null) {
+      node.put(columnName, bigDecimal);
+      return;
+    }
+
+    // ResultSet#getBigDecimal cannot handle Infinity, -Infinity, or NaN, and will throw exception,
+    // which becomes null. So we need to check these special values as string.
+    final String value = resultSet.getString(index);
+    if (value.equalsIgnoreCase("infinity")) {
+      node.put(columnName, Double.POSITIVE_INFINITY);
+    } else if (value.equalsIgnoreCase("-infinity")) {
+      node.put(columnName, Double.NEGATIVE_INFINITY);
+    } else if (value.equalsIgnoreCase("nan")) {
+      node.put(columnName, Double.NaN);
```

Comment:
does json have a concept of infinity and Nan? how are these represented when output from the connector itself?

Reply:
Good catch. Actually Json numbers do not support `NaN` or infinity. To support these three special values, we need to make sure that the destination can handle them. Otherwise, the destination will fail. Too bad, I need to revert this.

Reply:
like how we currently handle dates, you can output them in a string type with a format hint.

Then have normalization handle the special format hint for these special numbers accordingly for the destination

Reply:
@ChristopheDuong, nice.

Where the code that handles this? I'd like to see what the format hint looks like.

Reply:
In the json, it's "string" type.

Then, on normalization, if we find a format hint, we know it's actually a date not a string:
https://github.com/airbytehq/airbyte/blob/c5fc568061f00b99dac9ec3f0f1573e3a30c096a/airbyte-integrations/bases/base-normalization/normalization/transform_catalog/utils.py#L33

Reply:
Example of catalog.json with date:
https://github.com/airbytehq/airbyte/blob/c5fc568061f00b99dac9ec3f0f1573e3a30c096a/airbyte-integrations/bases/base-normalization/integration_tests/resources/test_simple_streams/data_input/catalog.json#L17

so you can have a special string-float type instead of using json's number in order to encode special float values:

```
"db_float_column": {
  "type": "string",
  "format": "airbyte-float"
}
```